File: src/shared/Network.lua
-- Shared Network Module
-- Centralizes access to RemoteEvents and provides a consistent API

local RunService = game:GetService("RunService")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Logger = require(ReplicatedStorage:WaitForChild("Shared"):WaitForChild("Logger"))

local Events = ReplicatedStorage:FindFirstChild("Events")
if not Events then
	if RunService:IsServer() then
		Events = Instance.new("Folder")
		Events.Name = "Events"
		Events.Parent = ReplicatedStorage
	else
		Events = ReplicatedStorage:WaitForChild("Events")
	end
end

local Network = {}

-- Utility to get a RemoteEvent
function Network.GetEvent(eventName)
	local event = Events:FindFirstChild(eventName)
	if not event then
		if RunService:IsServer() then
			-- Create it if it doesn't exist (Server side only)
			event = Instance.new("RemoteEvent")
			event.Name = eventName
			event.Parent = Events
			Logger.Debug("Network", "Created RemoteEvent: " .. eventName)
		else
			-- Client side: wait for it to exist
			event = Events:WaitForChild(eventName, 5)
			if not event then
				error("RemoteEvent not found: " .. eventName)
			end
		end
	end
	return event
end

-- Server -> Client: Fire to a specific player
function Network:FireClient(player, eventName, ...)
	if not RunService:IsServer() then
		error("FireClient can only be called from the server")
	end
	self.GetEvent(eventName):FireClient(player, ...)
end

-- Server -> Client: Fire to all players
function Network:FireAllClients(eventName, ...)
	if not RunService:IsServer() then
		error("FireAllClients can only be called from the server")
	end
	self.GetEvent(eventName):FireAllClients(...)
end

-- Client -> Server: Fire request
function Network:FireServer(eventName, ...)
	if not RunService:IsClient() then
		error("FireServer can only be called from the client")
	end
	-- We use a single ClientRequest event for all client-to-server requests
	-- The first argument is the specific action type
	self.GetEvent("ClientRequest"):FireServer(eventName, ...)
end

-- Connect to an event (Server or Client)
function Network:OnEvent(eventName, callback)
	local event = self.GetEvent(eventName)
	if RunService:IsServer() then
		return event.OnServerEvent:Connect(callback)
	else
		return event.OnClientEvent:Connect(callback)
	end
end

return Network

File: src/shared/Network.ts
// Shared Network Module
// Centralizes access to RemoteEvents and provides a consistent API

const RunService = game.GetService("RunService");
const ReplicatedStorage = game.GetService("ReplicatedStorage");
import * as Logger from "shared/Logger";

let events = ReplicatedStorage.FindFirstChild("Events");
if (!events) {
	if (RunService.IsServer()) {
		events = new Instance("Folder");
		events.Name = "Events";
		events.Parent = ReplicatedStorage;
	} else {
		events = ReplicatedStorage.WaitForChild("Events");
	}
}

const getEvent = (eventName: string) => {
	const eventsFolder = events as Folder;
	let event = eventsFolder.FindFirstChild(eventName);
	if (!event) {
		if (RunService.IsServer()) {
			// Create it if it doesn't exist (Server side only)
			event = new Instance("RemoteEvent");
			event.Name = eventName;
			event.Parent = eventsFolder;
			Logger.Debug("Network", `Created RemoteEvent: ${eventName}`);
		} else {
			// Client side: wait for it to exist
			const awaited = eventsFolder.WaitForChild(eventName, 5);
			if (!awaited || !awaited.IsA("RemoteEvent")) {
				throw `RemoteEvent not found: ${eventName}`;
			}
			event = awaited;
		}
	}

	if (!event.IsA("RemoteEvent")) {
		throw `RemoteEvent not found: ${eventName}`;
	}

	return event;
};

const FireClient = (player: Player, eventName: string, ...args: unknown[]) => {
	if (!RunService.IsServer()) {
		throw "FireClient can only be called from the server";
	}
	getEvent(eventName).FireClient(player, ...args);
};

const FireAllClients = (eventName: string, ...args: unknown[]) => {
	if (!RunService.IsServer()) {
		throw "FireAllClients can only be called from the server";
	}
	getEvent(eventName).FireAllClients(...args);
};

const FireServer = (eventName: string, ...args: unknown[]) => {
	if (!RunService.IsClient()) {
		throw "FireServer can only be called from the client";
	}
	// We use a single ClientRequest event for all client-to-server requests
	// The first argument is the specific action type
	getEvent("ClientRequest").FireServer(eventName, ...args);
};

const OnEvent = (eventName: string, callback: (...args: unknown[]) => void) => {
	const event = getEvent(eventName);
	if (RunService.IsServer()) {
		return event.OnServerEvent.Connect(callback);
	}
	return event.OnClientEvent.Connect(callback);
};

export default {
	GetEvent: getEvent,
	FireClient,
	FireAllClients,
	FireServer,
	OnEvent,
};
